lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs

```diff
@@ -215,28 +214,39 @@ spec = describe "Logging Middleware"
             , (Debug, "LogRequestFinish")
             ]
 
-    it "different request ids" $ \ctx -> property $ \(NumberOfRequests n) ->
-        monadicIO $ liftIO $ do
-            void $ mapConcurrently (const (get ctx "/get")) (replicate n ())
-            entries <- readTVarIO (logs ctx)
+    it "different request ids" $ \ctx ->
+        property $ \(NumberOfRequests n) -> monadicIO $ do
+            entries <- liftIO $ do
+                replicateConcurrently_ n (get ctx "/get")
+                skipPrevLogs <$> takeLogs ctx
             let getReqId (ApiLog (RequestId rid) _) = rid
             let uniqueReqIds = L.nubBy (\l1 l2 -> getReqId l1 == getReqId l2)
-            pendingOnWindows "Disabled on windows due to race with log flushing"
-            length (uniqueReqIds entries) `shouldBe` n
-
-    it "correct time measures" $ \ctx -> property $ \(nReq, ix) ->
-        withMaxSuccess 10 $ monadicIO $ liftIO $ do
-            let (NumberOfRequests n, RandomIndex i) = (nReq, ix)
-            let reqs = mconcat
-                    [ replicate i (get ctx "/get")
-                    , [ get ctx "/long" ]
-                    , replicate (n - i) (get ctx "/get")
-                    ]
-            void $ mapConcurrently id reqs
-            entries <- readTVarIO (logs ctx)
+            let numUniqueReqIds = length (uniqueReqIds entries)
+            monitor $ counterexample $ unlines $
+                [ "Number of log entries: " ++ show (length entries)
+                , "Number of unique req ids: " ++ show numUniqueReqIds
+                , ""
+                , "All the logs:" ] ++ map show entries
+            assert $ numUniqueReqIds == n
+
+    it "correct time measures" $ \ctx -> withMaxSuccess 10 $
+        property $ \(NumberOfRequests n, RandomIndex i) -> monadicIO $ do
+            entries <- liftIO $ do
+                let reqs = mconcat
+                        [ replicate i (get ctx "/get")
+                        , [ get ctx "/long" ]
+                        , replicate (n - i) (get ctx "/get")
+                        ]
+                void $ mapConcurrently id reqs
+                waitForServerToComplete
```

Comment:
Is this the only behavioral change which fixes the problem?

Reply:
This sleep fixes "correct time measures", which failed occasionally. Since there are only 10 examples repeated, it's ok to sleep a bit.

I also inserted waitForServerToComplete into expectLogs, because it might also be susceptible to the race condition.

With the "different request ids" property, which was the one which failed most often, I could not really use sleeps because there are 100 examples run. So in that test I fix it by filtering out any logs from the previous example.